ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp

```diff
@@ -284,10 +296,56 @@ static int HandleVnodeOperation(
                 }
             }
         }
+        else if (ActionBitIsSet(action, KAUTH_VNODE_DELETE))
+        {
+            if (!TrySendRequestAndWaitForResponse(
+                    root,
+                    MessageType_KtoU_NotifyDirectoryPreDelete,
+                    currentVnode,
+                    pid,
+                    procname,
+                    &kauthResult,
+                    kauthError))
+            {
+                goto CleanupAndReturn;
+            }
+        }
     }
     else
     {
-        if (ActionBitIsSet(
+        if (ActionBitIsSet(action, KAUTH_VNODE_DELETE))
+        {
+            if (!TrySendRequestAndWaitForResponse(
+                    root,
+                    MessageType_KtoU_NotifyFilePreDelete,
+                    currentVnode,
+                    pid,
+                    procname,
+                    &kauthResult,
+                    kauthError))
+            {
+                goto CleanupAndReturn;
+            }
+            
+            // This delete could be part of a rename, and so we must hydrate the file now to
+            // ensure it's hydrated post-rename
+            if (kauthResult == KAUTH_RESULT_DEFER &&
+                FileFlagsBitIsSet(currentVnodeFileFlags, FileFlags_IsEmpty))
+            {
+                if (!TrySendRequestAndWaitForResponse(
```

Comment:
Can we avoid duplicating this logic for hydrating a file? It seems like `KAUTH_VNODE_DELETE` should just be another action that can trigger a hydration in the existing code path. Then after that, we could check for `KAUTH_VNODE_DELETE` again and trigger the pre-delete notification.

Reply:
I think we could move the PreDelete notification code down and avoid duplicating the logic.  It's a trade-off between avoiding duplicating the logic and preventing an unnecessary hydration if a provider blocks the delete.

However, blocking the delete of an unhydrated would be quite rare (and it's just a perf tradeoff), and so I'll go with your suggestion as I'd rather keep the kext simpler at this point.